src/metadata_storage_stb_ds.cc

```diff
@@ -272,23 +272,111 @@ void AllocateOrGrowIdList(MetadataManager *mdm, ChunkedIdList *id_list) {
+                            u32 index) {
+  u64 result = 0;
+  if (id_list->length && index < id_list->length) {
+    u64 *head = GetIdsPtr(mdm, *id_list);
+    result = head[index];
+    ReleaseIdsPtr(mdm);
+  }
+
+  return result;
+}
+
+void SetChunkedIdListElement(MetadataManager *mdm, ChunkedIdList *id_list,
+                             u32 index, u64 value) {
+  if (id_list->length && index >= id_list->length) {
+    LOG(WARNING) << "Attempting to set index " << index
+                 << " on a ChunkedIdList of length " << id_list->length
+                 << std::endl;
+  } else {
+    u64 *head = GetIdsPtr(mdm, *id_list);
+    head[index] = value;
+    ReleaseIdsPtr(mdm);
+  }
+}
+
+void LocalIncrementBlobStats(MetadataManager *mdm, ChunkedIdList *stats,
+                             u32 index) {
+  Stats stat = {};
+  stat.as_int = GetChunkedIdListElement(mdm, stats, index);
+  stat.bits.frequency++;
+  stat.bits.recency = mdm->clock++;
+
+  SetChunkedIdListElement(mdm, stats, index, stat.as_int);
+}
+
+i64 GetIndexOfId(MetadataManager *mdm, ChunkedIdList *id_list, u64 id) {
+  i64 result = -1;
+
+  u64 *head = GetIdsPtr(mdm, *id_list);
+  for (i64 i = 0; i < id_list->length; ++i) {
+    if (head[i] == id) {
+      result = i;
+      break;
+    }
+  }
+  ReleaseIdsPtr(mdm);
+
+  return result;
+}
+
+/** Protects the @p bucket_id's 'blobs' ChunkedIdList before incrementing its
+ *  Stats.
+ */
+void LocalIncrementBlobStatsSafely(MetadataManager *mdm, BucketID bucket_id,
+                                   BlobID blob_id) {
+  BeginTicketMutex(&mdm->bucket_mutex);
+  BucketInfo *info = LocalGetBucketInfoById(mdm, bucket_id);
+  i64 index = GetIndexOfId(mdm, &info->blobs, blob_id.as_int);
+  if (index < 0) {
+    LOG(WARNING) << "BlobID " << blob_id.as_int << " not found in Bucket "
+                 << bucket_id.as_int << std::endl;
+  } else {
+    LocalIncrementBlobStats(mdm, &info->stats, (u32)index);
+  }
+  EndTicketMutex(&mdm->bucket_mutex);
+}
+
+void IncrementBlobStatsSafely(SharedMemoryContext *context, RpcContext *rpc,
+                              const std::string &name, BucketID bucket_id,
+                              BlobID blob_id) {
+  MetadataManager *mdm = GetMetadataManagerFromContext(context);
+  u32 target_node = HashString(mdm, rpc, name.c_str());
+
+  if (target_node == rpc->node_id) {
+    LocalIncrementBlobStatsSafely(mdm, bucket_id, blob_id);
+  } else {
+    RpcCall<bool>(rpc, target_node, "RemoteIncrementBlobStatsSafely", bucket_id,
+                  blob_id);
+  }
 }
 
 void LocalAddBlobIdToBucket(MetadataManager *mdm, BucketID bucket_id,
                             BlobID blob_id) {
   BeginTicketMutex(&mdm->bucket_mutex);
   BucketInfo *info = LocalGetBucketInfoById(mdm, bucket_id);
-  AppendToChunkedIdList(mdm, &info->blobs, blob_id.as_int);
+  u32 index = AppendToChunkedIdList(mdm, &info->blobs, blob_id.as_int);
+  AppendToChunkedIdList(mdm, &info->stats, 0);
```

Comment:
What is the reason to call `AppendToChunkedIdList` twice with different last parameter?

Reply:
The first call is to add the `BlobID` to the `Bucket`'s list of blobs. The second call is to add the blob statistics, which I treat as an ID here because it can be read as a `u64`. `info->stats[i]` represents the stats for `info->blobs[i]`.This will go away once I implement #213.